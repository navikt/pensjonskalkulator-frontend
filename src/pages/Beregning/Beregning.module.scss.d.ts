declare const classNames: typeof globalClassNames & {
  readonly beregning: "beregning";
  readonly container: "container";
  readonly container__hasPadding: "container__hasPadding";
  readonly alert: "alert";
  readonly toggle: "toggle";
  readonly toggle__paddingBottom: "toggle__paddingBottom";
  readonly background: "background";
  readonly background__lightblue: "background__lightblue";
};
export = classNames;
