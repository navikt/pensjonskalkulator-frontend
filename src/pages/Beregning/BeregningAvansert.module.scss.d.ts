declare const classNames: typeof globalClassNames & {
  readonly container: "container";
  readonly container__hasMobilePadding: "container__hasMobilePadding";
  readonly container__hasTopMargin: "container__hasTopMargin";
  readonly background: "background";
  readonly background__lightblue: "background__lightblue";
  readonly button: "button";
};
export = classNames;
