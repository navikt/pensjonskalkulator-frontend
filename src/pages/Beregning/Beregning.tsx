import clsx from 'clsx'
import React from 'react'
import { FormattedMessage, useIntl } from 'react-intl'
import { useNavigate } from 'react-router'

import { Alert, Button, Modal, ToggleGroup } from '@navikt/ds-react'

import { InfoOmFremtidigVedtak } from '@/components/InfoOmFremtidigVedtak'
import { LightBlueFooter } from '@/components/LightBlueFooter'
import { ShowMoreRef } from '@/components/common/ShowMore/ShowMore'
import { paths } from '@/router/constants'
import { useAppDispatch, useAppSelector } from '@/state/hooks'
import { selectHasErApotekerError } from '@/state/session/selectors'
import {
  selectCurrentSimulation,
  selectFoedselsdato,
  selectIsEndring,
  selectLoependeVedtak,
  selectSkalBeregneAfpKap19,
} from '@/state/userInput/selectors'
import { userInputActions } from '@/state/userInput/userInputSlice'
import { BeregningVisning } from '@/types/common-types'
import { isFoedtEtter1963 } from '@/utils/alder'
import { BUTTON_KLIKK, MODAL_AAPNET } from '@/utils/loggerConstants'
import { logger } from '@/utils/logging'

import { BeregningAvansert } from './BeregningAvansert'
import { BeregningEnkel } from './BeregningEnkel'
import { AvansertBeregningModus, BeregningContext } from './context'

import styles from './Beregning.module.scss'

interface Props {
  visning: BeregningVisning
}

export const Beregning: React.FC<Props> = ({ visning }) => {
  const intl = useIntl()
  const navigate = useNavigate()
  const dispatch = useAppDispatch()

  const APPLICATION_TITLE_BEREGNING = 'application.title.beregning'

  const { uttaksalder } = useAppSelector(selectCurrentSimulation)
  const avbrytModalRef = React.useRef<HTMLDialogElement>(null)

  const [avansertSkjemaModus, setAvansertSkjemaModus] =
    React.useState<AvansertBeregningModus>('redigering')
  const [harAvansertSkjemaUnsavedChanges, setHarAvansertSkjemaUnsavedChanges] =
    React.useState<boolean>(false)

  const isEndring = useAppSelector(selectIsEndring)
  const loependeVedtak = useAppSelector(selectLoependeVedtak)
  const skalBeregneAfpKap19 = useAppSelector(selectSkalBeregneAfpKap19)
  const foedselsdato = useAppSelector(selectFoedselsdato)
  const foedtEtter1963 = isFoedtEtter1963(foedselsdato)
  const hasErApotekerError = useAppSelector(selectHasErApotekerError)

  React.useEffect(() => {
    document.title = intl.formatMessage({
      id: APPLICATION_TITLE_BEREGNING,
    })
  }, [])

  const shouldShowModalBoolean = React.useMemo(() => {
    return (
      harAvansertSkjemaUnsavedChanges ||
      avansertSkjemaModus === 'resultat' ||
      !!(avansertSkjemaModus === 'redigering' && uttaksalder)
    )
  }, [uttaksalder, avansertSkjemaModus, harAvansertSkjemaUnsavedChanges])

  React.useEffect(() => {
    let isEventAdded
    const onPopState = () => {
      logger(MODAL_AAPNET, {
        tekst: 'Modal: Er du sikker på at du vil avslutte avansert beregning?',
      })
      avbrytModalRef.current?.showModal()
    }

    if (window.location.href.includes(paths.beregningAvansert)) {
      if (shouldShowModalBoolean) {
        window.history.pushState(
          null,
          intl.formatMessage({
            id: APPLICATION_TITLE_BEREGNING,
          }),
          window.location.href
        )
        window.addEventListener('popstate', onPopState)
        isEventAdded = true
      } else {
        window.removeEventListener('popstate', onPopState)
        if (isEventAdded) {
          isEventAdded = false
          navigate(-1)
        }
      }
    }

    return () => {
      if (onPopState) {
        window.removeEventListener('popstate', onPopState)
        isEventAdded = false
      }
    }
  }, [shouldShowModalBoolean])

  const navigateToTab = (v: BeregningVisning) => {
    navigate(v === 'enkel' ? paths.beregningEnkel : paths.beregningAvansert)
    dispatch(userInputActions.flushCurrentSimulation())
    setAvansertSkjemaModus('redigering')
    setHarAvansertSkjemaUnsavedChanges(false)
  }

  const onToggleChange = (v: string) => {
    logger(BUTTON_KLIKK, {
      tekst: `Toggle viser fane ${v}`,
    })
    if (
      visning === 'avansert' &&
      v === 'enkel' &&
      (harAvansertSkjemaUnsavedChanges ||
        avansertSkjemaModus === 'resultat' ||
        (avansertSkjemaModus === 'redigering' && uttaksalder))
    ) {
      logger(MODAL_AAPNET, {
        tekst: 'Modal: Er du sikker på at du vil avslutte avansert beregning?',
      })
      avbrytModalRef.current?.showModal()
    } else {
      navigateToTab(v as BeregningVisning)
    }
  }

  const pensjonsavtalerShowMoreRef = React.useRef<ShowMoreRef>(null)

  return (
    <BeregningContext.Provider
      value={{
        avansertSkjemaModus,
        setAvansertSkjemaModus,
        harAvansertSkjemaUnsavedChanges,
        setHarAvansertSkjemaUnsavedChanges,
        pensjonsavtalerShowMoreRef,
      }}
    >
      <Modal
        ref={avbrytModalRef}
        header={{
          heading: intl.formatMessage({
            id: isEndring
              ? 'beregning.avansert.avbryt_modal.endring.title'
              : 'beregning.avansert.avbryt_modal.title',
          }),
        }}
        width="medium"
        onClose={() => {
          if (window.location.href.includes(paths.beregningAvansert)) {
            window.history.pushState(
              null,
              intl.formatMessage({
                id: APPLICATION_TITLE_BEREGNING,
              }),
              window.location.href
            )
          }
        }}
      >
        <Modal.Footer>
          <Button
            type="button"
            onClick={() => {
              if (isEndring) {
                dispatch(userInputActions.flushCurrentSimulation())
                navigate(paths.start)
              } else {
                navigateToTab('enkel')
              }

              avbrytModalRef.current?.close('returnValue')
            }}
          >
            {intl.formatMessage({
              id: isEndring
                ? 'beregning.avansert.avbryt_modal.endring.button.avslutt'
                : 'beregning.avansert.avbryt_modal.button.avslutt',
            })}
          </Button>
          <Button
            type="button"
            variant="secondary"
            onClick={() => {
              avbrytModalRef.current?.close()
            }}
          >
            {intl.formatMessage({
              id: 'beregning.avansert.avbryt_modal.button.avbryt',
            })}
          </Button>
        </Modal.Footer>
      </Modal>

      <div className={styles.beregning}>
        <div className={styles.container}>
          <InfoOmFremtidigVedtak loependeVedtak={loependeVedtak} />
        </div>

        {hasErApotekerError && foedtEtter1963 && (
          <div className={styles.container}>
            <Alert
              className={styles.alert}
              variant="warning"
              aria-live="polite"
            >
              <FormattedMessage id="error.apoteker_warning" />
            </Alert>
          </div>
        )}

        {!isEndring && !skalBeregneAfpKap19 && (
          <div className={styles.toggle}>
            <div className={styles.container} data-testid="toggle-avansert">
              <ToggleGroup
                value={visning}
                variant="neutral"
                onChange={onToggleChange}
              >
                <ToggleGroup.Item value="enkel">
                  {intl.formatMessage({
                    id: 'beregning.toggle.enkel',
                  })}
                </ToggleGroup.Item>

                <ToggleGroup.Item value="avansert">
                  {intl.formatMessage({
                    id: 'beregning.toggle.avansert',
                  })}
                </ToggleGroup.Item>
              </ToggleGroup>
            </div>
          </div>
        )}

        {visning === 'enkel' && <BeregningEnkel />}

        {visning === 'avansert' && <BeregningAvansert />}

        <div className={clsx(styles.background, styles.background__lightblue)}>
          <div className={styles.container}>
            <LightBlueFooter />
          </div>
        </div>
      </div>
    </BeregningContext.Provider>
  )
}
