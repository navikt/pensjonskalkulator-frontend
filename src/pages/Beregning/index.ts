export { Beregning } from './Beregning'
export { BeregningAvansert } from './BeregningAvansert'
