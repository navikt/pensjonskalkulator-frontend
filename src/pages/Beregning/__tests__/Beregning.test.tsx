import { describe, expect, it, vi } from 'vitest'

import { Beregning } from '../Beregning'
import { AVANSERT_FORM_NAMES } from '@/components/AvansertSkjema/utils'
import {
  fulfilledGetInntekt,
  fulfilledGetPerson,
  fulfilledGetLoependeVedtak0Ufoeregrad,
  fulfilledGetLoependeVedtakLoependeAlderspensjon,
  fulfilledGetLoependeVedtakFremtidig,
  fulfilledGetLoependeVedtakFremtidigMedAlderspensjon,
} from '@/mocks/mockedRTKQueryApiCalls'
import { paths } from '@/router/constants'
import { userInputInitialState } from '@/state/userInput/userInputReducer'
import * as userInputReducerUtils from '@/state/userInput/userInputReducer'
import { fireEvent, render, screen, userEvent, waitFor } from '@/test-utils'
const previousWindow = window

const navigateMock = vi.fn()
vi.mock(import('react-router'), async (importOriginal) => {
  const actual = await importOriginal()
  return {
    ...actual,
    useNavigate: () => navigateMock,
  }
})

describe('Beregning', () => {
  const preloadedQueries = {
    api: {
      queries: {
        ...fulfilledGetPerson,
        ...fulfilledGetInntekt,
        ...fulfilledGetLoependeVedtak0Ufoeregrad,
      },
    },
  }

  it('har riktig sidetittel', () => {
    render(<Beregning visning="enkel" />, {
      preloadedState: {
        api: {
          // @ts-ignore
          queries: {
            ...fulfilledGetPerson,
            ...fulfilledGetInntekt,
            ...fulfilledGetLoependeVedtakLoependeAlderspensjon,
          },
        },
        userInput: {
          ...userInputInitialState,
          samtykke: false,
          currentSimulation: {
            ...userInputInitialState.currentSimulation,
            formatertUttaksalderReadOnly:
              '70 alder.aar string.og 4 alder.maaned',
            uttaksalder: { aar: 70, maaneder: 4 },
          },
        },
      },
    })
    expect(document.title).toBe('application.title.beregning')
  })

  describe('Gitt at brukeren har vedtak om alderspensjon', () => {
    it('viser ikke toggle på toppen av siden', async () => {
      render(<Beregning visning="enkel" />, {
        preloadedState: {
          api: {
            // @ts-ignore
            queries: {
              ...fulfilledGetPerson,
              ...fulfilledGetInntekt,
              ...fulfilledGetLoependeVedtakLoependeAlderspensjon,
            },
          },
          userInput: {
            ...userInputInitialState,
            samtykke: false,
            currentSimulation: {
              utenlandsperioder: [],
              formatertUttaksalderReadOnly:
                '70 alder.aar string.og 4 alder.maaned',
              uttaksalder: { aar: 70, maaneder: 4 },
              aarligInntektFoerUttakBeloep: '300 000',
              gradertUttaksperiode: null,
            },
          },
        },
      })

      expect(screen.queryByTestId('toggle-avansert')).not.toBeInTheDocument()
    })

    it('når vedtaket gjelder frem i tid vises info om det på toppen av siden', async () => {
      render(<Beregning visning="enkel" />, {
        preloadedState: {
          api: {
            // @ts-ignore
            queries: {
              ...fulfilledGetPerson,
              ...fulfilledGetInntekt,
              ...fulfilledGetLoependeVedtakFremtidig,
            },
          },
          userInput: {
            ...userInputInitialState,
            samtykke: false,
            currentSimulation: {
              utenlandsperioder: [],
              formatertUttaksalderReadOnly:
                '70 alder.aar string.og 4 alder.maaned',
              uttaksalder: { aar: 70, maaneder: 4 },
              aarligInntektFoerUttakBeloep: '300 000',
              gradertUttaksperiode: null,
            },
          },
        },
      })

      expect(
        screen.getByText('stegvisning.fremtidigvedtak.alert')
      ).toBeVisible()
    })

    it('når vedtaket gjelder både nå og frem i tid vises info om det på toppen av siden', async () => {
      render(<Beregning visning="enkel" />, {
        preloadedState: {
          api: {
            // @ts-ignore
            queries: {
              ...fulfilledGetPerson,
              ...fulfilledGetInntekt,
              ...fulfilledGetLoependeVedtakFremtidigMedAlderspensjon,
            },
          },
          userInput: {
            ...userInputInitialState,
            samtykke: false,
            currentSimulation: {
              utenlandsperioder: [],
              formatertUttaksalderReadOnly:
                '70 alder.aar string.og 4 alder.maaned',
              uttaksalder: { aar: 70, maaneder: 4 },
              aarligInntektFoerUttakBeloep: '300 000',
              gradertUttaksperiode: null,
            },
          },
        },
      })

      expect(
        screen.getByText('stegvisning.fremtidigvedtak.endring.alert')
      ).toBeVisible()
    })
  })

  describe('Gitt at brukeren navigerer mellom fanene', () => {
    it('når brukeren har gjort en Enkel simulering og bytter fane, nullstiller det pågående simulering', async () => {
      const user = userEvent.setup()
      const flushCurrentSimulationMock = vi.spyOn(
        userInputReducerUtils.userInputActions,
        'flushCurrentSimulationUtenomUtenlandsperioder'
      )
      render(<Beregning visning="enkel" />, {
        preloadedState: {
          // @ts-ignore
          api: {
            ...preloadedQueries.api,
          },
          userInput: {
            ...userInputInitialState,
            samtykke: true,
            currentSimulation: {
              utenlandsperioder: [],
              formatertUttaksalderReadOnly:
                '70 alder.aar string.og 4 alder.maaned',
              uttaksalder: { aar: 70, maaneder: 4 },
              aarligInntektFoerUttakBeloep: '300 000',
              gradertUttaksperiode: null,
            },
          },
        },
      })

      expect(await screen.findByTestId('toggle-avansert')).toBeVisible()
      await user.click(await screen.findByText('beregning.toggle.avansert'))
      expect(flushCurrentSimulationMock).toHaveBeenCalled()
    })

    it('når brukeren begynner å fylle ut skjema på Avansert og bytter fane, gir Modalen muligheten til å avbryte eller avslutte beregningen', async () => {
      const user = userEvent.setup()
      const flushCurrentSimulationMock = vi.spyOn(
        userInputReducerUtils.userInputActions,
        'flushCurrentSimulationUtenomUtenlandsperioder'
      )
      render(<Beregning visning="avansert" />, {
        preloadedState: {
          // @ts-ignore
          api: {
            ...preloadedQueries.api,
          },
          userInput: {
            ...userInputInitialState,
          },
        },
      })

      fireEvent.change(
        await screen.findByTestId(
          `age-picker-${AVANSERT_FORM_NAMES.uttaksalderHeltUttak}-aar`
        ),
        {
          target: { value: '67' },
        }
      )

      expect(await screen.findByTestId('toggle-avansert')).toBeVisible()
      await user.click(await screen.findByText('beregning.toggle.enkel'))
      expect(
        await screen.findByText('beregning.avansert.avbryt_modal.title')
      ).toBeVisible()
      await user.click(
        await screen.findByText('beregning.avansert.avbryt_modal.button.avbryt')
      )
      await user.click(await screen.findByText('beregning.toggle.enkel'))
      await user.click(
        await screen.findByText(
          'beregning.avansert.avbryt_modal.button.avslutt'
        )
      )
      expect(navigateMock).toHaveBeenCalledWith(paths.beregningEnkel)
      expect(flushCurrentSimulationMock).toHaveBeenCalled()
    })

    it('når brukeren er på resultatside etter en Avansert simulering og bytter fane, gir Modalen muligheten til å avbryte eller avslutte beregningen', async () => {
      const user = userEvent.setup()
      render(<Beregning visning="avansert" />, {
        preloadedState: {
          // @ts-ignore
          api: {
            ...preloadedQueries.api,
          },
          userInput: {
            ...userInputInitialState,
          },
        },
      })

      fireEvent.change(
        await screen.findByTestId(
          `age-picker-${AVANSERT_FORM_NAMES.uttaksalderHeltUttak}-aar`
        ),
        {
          target: { value: '67' },
        }
      )
      fireEvent.change(
        await screen.findByTestId(
          `age-picker-${AVANSERT_FORM_NAMES.uttaksalderHeltUttak}-maaneder`
        ),
        {
          target: { value: '0' },
        }
      )

      await user.click(
        await screen.findByText('beregning.avansert.button.beregn')
      )
      expect(await screen.findByTestId('toggle-avansert')).toBeVisible()
      await user.click(await screen.findByText('beregning.toggle.enkel'))
      expect(
        await screen.findByText('beregning.avansert.avbryt_modal.title')
      ).toBeVisible()
      expect(
        await screen.findByText('beregning.avansert.avbryt_modal.button.avbryt')
      ).toBeVisible()
      expect(
        await screen.findByText(
          'beregning.avansert.avbryt_modal.button.avslutt'
        )
      ).toBeVisible()
    })

    it('når brukeren har gjort en Avansert simulering som hen redigerer og bytter fane, gir Modalen muligheten til å avbryte eller avslutte beregningen', async () => {
      const user = userEvent.setup()
      render(<Beregning visning="avansert" />, {
        preloadedState: {
          // @ts-ignore
          api: {
            ...preloadedQueries.api,
          },
          userInput: {
            ...userInputInitialState,
            samtykke: false,
            currentSimulation: {
              ...userInputInitialState.currentSimulation,
              formatertUttaksalderReadOnly:
                '70 alder.aar string.og 4 alder.maaned',
              uttaksalder: { aar: 70, maaneder: 4 },
            },
          },
        },
      })

      expect(await screen.findByTestId('toggle-avansert')).toBeVisible()
      await user.click(await screen.findByText('beregning.toggle.enkel'))
      expect(
        await screen.findByText('beregning.avansert.avbryt_modal.title')
      ).toBeVisible()
      expect(
        await screen.findByText('beregning.avansert.avbryt_modal.button.avbryt')
      ).toBeVisible()
      expect(
        await screen.findByText(
          'beregning.avansert.avbryt_modal.button.avslutt'
        )
      ).toBeVisible()
    })
  })

  describe('Gitt at brukeren navigerer tilbake', () => {
    beforeEach(() => {
      global.window = Object.create(window)
      Object.defineProperty(window, 'location', {
        value: {
          href: paths.beregningAvansert,
        },
        writable: true,
      })
    })

    afterEach(() => {
      global.window = previousWindow
    })

    function NavigateWrapper({ children }: { children: React.ReactNode }) {
      function handleClick() {
        fireEvent(
          window,
          new window.PopStateEvent('popstate', {
            state: { page: 1 },
          })
        )
      }

      return (
        <div>
          <button onClick={handleClick} data-testid="navigate-btn">
            Navigate
          </button>
          {children}
        </div>
      )
    }

    it('når brukeren har gjort en Enkel simulering og trykker på tilbakeknappen, vises ikke Avbryt-Modalen', async () => {
      global.window = Object.create(window)
      Object.defineProperty(window, 'location', {
        value: {
          href: paths.beregningEnkel,
        },
        writable: true,
      })
      const user = userEvent.setup()
      render(
        <NavigateWrapper>
          <Beregning visning="enkel" />
        </NavigateWrapper>,
        {
          preloadedState: {
            // @ts-ignore
            api: {
              ...preloadedQueries.api,
            },
            userInput: {
              ...userInputInitialState,
              samtykke: true,
              currentSimulation: {
                utenlandsperioder: [],
                formatertUttaksalderReadOnly:
                  '70 alder.aar string.og 4 alder.maaned',
                uttaksalder: { aar: 70, maaneder: 4 },
                aarligInntektFoerUttakBeloep: '300 000',
                gradertUttaksperiode: null,
              },
            },
          },
        }
      )

      expect(await screen.findByTestId('toggle-avansert')).toBeInTheDocument()
      await user.click(await screen.findByTestId('navigate-btn'))
      expect(
        await screen.findByText('beregning.avansert.avbryt_modal.title')
      ).not.toBeVisible()
    })

    it('når brukeren begynner å fylle ut skjema på Avansert og trykker på tilbakeknappen, vises Avbryt-Modalen', async () => {
      const user = userEvent.setup()

      render(
        <NavigateWrapper>
          <Beregning visning="avansert" />
        </NavigateWrapper>,
        {
          preloadedState: {
            // @ts-ignore
            api: {
              ...preloadedQueries.api,
            },
            userInput: {
              ...userInputInitialState,
            },
          },
        }
      )

      fireEvent.change(
        await screen.findByTestId(
          `age-picker-${AVANSERT_FORM_NAMES.uttaksalderHeltUttak}-aar`
        ),
        {
          target: { value: '67' },
        }
      )

      expect(await screen.findByTestId('toggle-avansert')).toBeInTheDocument()
      await user.click(await screen.findByTestId('navigate-btn'))
      expect(
        await screen.findByText('beregning.avansert.avbryt_modal.title')
      ).toBeVisible()
    })

    it('når brukeren er på resultatside etter en Avansert simulering og trykker på tilbakeknappen, vises Avbryt-Modalen', async () => {
      const user = userEvent.setup()
      render(
        <NavigateWrapper>
          <Beregning visning="avansert" />
        </NavigateWrapper>,
        {
          preloadedState: {
            // @ts-ignore
            api: {
              ...preloadedQueries.api,
            },
            userInput: {
              ...userInputInitialState,
            },
          },
        }
      )

      fireEvent.change(
        await screen.findByTestId(
          `age-picker-${AVANSERT_FORM_NAMES.uttaksalderHeltUttak}-aar`
        ),
        {
          target: { value: '67' },
        }
      )
      fireEvent.change(
        await screen.findByTestId(
          `age-picker-${AVANSERT_FORM_NAMES.uttaksalderHeltUttak}-maaneder`
        ),
        {
          target: { value: '0' },
        }
      )

      await user.click(
        await screen.findByText('beregning.avansert.button.beregn')
      )
      expect(await screen.findByTestId('toggle-avansert')).toBeInTheDocument()
      await user.click(await screen.findByTestId('navigate-btn'))
      expect(
        await screen.findByText('beregning.avansert.avbryt_modal.title')
      ).toBeVisible()
    })

    it('når brukeren har gjort en Avansert simulering og trykker på tilbakeknappen, vises Avbryt-Modalen', async () => {
      const user = userEvent.setup()
      render(
        <NavigateWrapper>
          <Beregning visning="avansert" />
        </NavigateWrapper>,
        {
          preloadedState: {
            // @ts-ignore
            api: {
              ...preloadedQueries.api,
            },
            userInput: {
              ...userInputInitialState,
              samtykke: false,
              currentSimulation: {
                ...userInputInitialState.currentSimulation,
                formatertUttaksalderReadOnly:
                  '70 alder.aar string.og 4 alder.maaned',
                uttaksalder: { aar: 70, maaneder: 4 },
              },
            },
          },
        }
      )

      expect(await screen.findByTestId('toggle-avansert')).toBeInTheDocument()
      await user.click(await screen.findByTestId('navigate-btn'))
      expect(
        await screen.findByText('beregning.avansert.avbryt_modal.title')
      ).toBeVisible()
    })

    it('når brukeren med vedtak om alderspensjon er på resultatside etter en Avansert simulering og trykker på tilbakeknappen, vises Avbryt-Modalen og brukeren sendes til /start ved bekreftelse', async () => {
      const user = userEvent.setup()
      const flushCurrentSimulationMock = vi.spyOn(
        userInputReducerUtils.userInputActions,
        'flushCurrentSimulationUtenomUtenlandsperioder'
      )
      render(
        <NavigateWrapper>
          <Beregning visning="avansert" />
        </NavigateWrapper>,
        {
          preloadedState: {
            api: {
              // @ts-ignore
              queries: {
                ...fulfilledGetPerson,
                ...fulfilledGetInntekt,
                ...fulfilledGetLoependeVedtakLoependeAlderspensjon,
              },
            },
            userInput: {
              ...userInputInitialState,
            },
          },
        }
      )

      fireEvent.change(
        await screen.findByTestId(
          `age-picker-${AVANSERT_FORM_NAMES.uttaksalderHeltUttak}-aar`
        ),
        {
          target: { value: '67' },
        }
      )
      fireEvent.change(
        await screen.findByTestId(
          `age-picker-${AVANSERT_FORM_NAMES.uttaksalderHeltUttak}-maaneder`
        ),
        {
          target: { value: '0' },
        }
      )
      fireEvent.change(
        await screen.findByTestId(AVANSERT_FORM_NAMES.uttaksgrad),
        {
          target: { value: '100 %' },
        }
      )
      const heltRadioGroup = screen.getByTestId(
        AVANSERT_FORM_NAMES.inntektVsaHeltUttakRadio
      )
      expect(heltRadioGroup).toBeVisible()
      await user.click(
        screen.getByTestId(
          `${AVANSERT_FORM_NAMES.inntektVsaHeltUttakRadio}-nei`
        )
      )

      await user.click(
        screen.getByText(`beregning.avansert.button.beregn.endring`)
      )

      await user.click(await screen.findByTestId('navigate-btn'))

      expect(
        await screen.findByText('beregning.avansert.avbryt_modal.endring.title')
      ).toBeVisible()
      expect(
        screen.queryByText('beregning.avansert.avbryt_modal.title')
      ).not.toBeInTheDocument()
      expect(
        screen.queryByText('beregning.avansert.avbryt_modal.button.avslutt')
      ).not.toBeInTheDocument()
      await user.click(
        await screen.findByText('beregning.avansert.avbryt_modal.button.avbryt')
      )
      await user.click(await screen.findByTestId('navigate-btn'))
      await user.click(
        await screen.findByText(
          'beregning.avansert.avbryt_modal.endring.button.avslutt'
        )
      )
      expect(navigateMock).toHaveBeenCalledWith(paths.start)
      expect(flushCurrentSimulationMock).toHaveBeenCalled()
    })
  })

  describe('Gitt at pensjonskalkulator er i "enkel" visning', () => {
    it('vises det riktig innhold', async () => {
      render(<Beregning visning="enkel" />, {
        preloadedState: {
          // @ts-ignore
          api: {
            ...preloadedQueries.api,
          },
          userInput: {
            ...userInputInitialState,
          },
        },
      })
      expect(screen.getByTestId('uttaksalder-loader')).toBeVisible()
      await waitFor(async () => {
        expect(
          screen.queryByTestId('uttaksalder-loader')
        ).not.toBeInTheDocument()
      })
      expect(await screen.findByTestId('tidligst-mulig-uttak')).toBeVisible()
    })
  })

  describe('Gitt at pensjonskalkulator er i "avansert" visning', () => {
    it('vises det riktig innhold', async () => {
      render(<Beregning visning="avansert" />, {
        preloadedState: {
          // @ts-ignore
          api: {
            ...preloadedQueries.api,
          },
          userInput: {
            ...userInputInitialState,
          },
        },
      })
      expect(
        await screen.findByText(
          'beregning.avansert.rediger.inntekt_frem_til_uttak.label'
        )
      ).toBeInTheDocument()
    })
  })

  it('gir mulighet til å avbryte og starte ny beregning ', async () => {
    render(<Beregning visning="enkel" />, {
      preloadedState: {
        // @ts-ignore
        api: {
          ...preloadedQueries.api,
        },
        userInput: {
          ...userInputInitialState,
        },
      },
    })
    expect(await screen.findByText('stegvisning.tilbake_start')).toBeVisible()
  })
})
