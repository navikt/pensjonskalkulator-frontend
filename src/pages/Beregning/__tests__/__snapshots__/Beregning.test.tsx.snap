// Vitest Snapshot v1

exports[`Beregning > Når brukeren velger uttaksalder > viser feilmelding og skjuler Grunnlag når tidligste-uttaksalder har feilet og brukeren prøver å simulere med for lav uttaksalder 1`] = `
<DocumentFragment>
  <div
    class="_background_21eff7 _background__hasMargin_21eff7 _background__white_21eff7"
  >
    <div
      class="_container_21eff7"
    >
      <div
        class="_wrapper_4ca3e3"
      >
        <div
          class="_wrapperCard_4ca3e3"
        >
          <span
            class="_pin_4ca3e3"
          />
          <h2
            class="navds-heading navds-heading--xsmall"
          >
            Når vil du ta ut alderspensjon?
          </h2>
          <ul
            class="navds-chips _chipsWrapper_4ca3e3 _chipsWrapper__hasGap_4ca3e3 navds-chips--medium navds-body-short navds-body-short--small"
          >
            <li>
              <button
                aria-pressed="true"
                class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
              >
                <span
                  class="navds-chips__chip-text"
                >
                  62 år
                </span>
              </button>
            </li>
            <li>
              <button
                aria-pressed="false"
                class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
              >
                <span
                  class="navds-chips__chip-text"
                >
                  63 år
                </span>
              </button>
            </li>
            <li>
              <button
                aria-pressed="false"
                class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
              >
                <span
                  class="navds-chips__chip-text"
                >
                  64 år
                </span>
              </button>
            </li>
            <li>
              <button
                aria-pressed="false"
                class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
              >
                <span
                  class="navds-chips__chip-text"
                >
                  65 år
                </span>
              </button>
            </li>
            <li>
              <button
                aria-pressed="false"
                class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
              >
                <span
                  class="navds-chips__chip-text"
                >
                  66 år
                </span>
              </button>
            </li>
            <li>
              <button
                aria-pressed="false"
                class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
              >
                <span
                  class="navds-chips__chip-text"
                >
                  67 år
                </span>
              </button>
            </li>
            <li>
              <button
                aria-pressed="false"
                class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
              >
                <span
                  class="navds-chips__chip-text"
                >
                  68 år
                </span>
              </button>
            </li>
            <li>
              <button
                aria-pressed="false"
                class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
              >
                <span
                  class="navds-chips__chip-text"
                >
                  69 år
                </span>
              </button>
            </li>
            <li>
              <button
                aria-pressed="false"
                class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
              >
                <span
                  class="navds-chips__chip-text"
                >
                  70 år
                </span>
              </button>
            </li>
            <li>
              <button
                aria-pressed="false"
                class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
              >
                <span
                  class="navds-chips__chip-text"
                >
                  71 år
                </span>
              </button>
            </li>
            <li>
              <button
                aria-pressed="false"
                class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
              >
                <span
                  class="navds-chips__chip-text"
                >
                  72 år
                </span>
              </button>
            </li>
            <li>
              <button
                aria-pressed="false"
                class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
              >
                <span
                  class="navds-chips__chip-text"
                >
                  73 år
                </span>
              </button>
            </li>
            <li>
              <button
                aria-pressed="false"
                class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
              >
                <span
                  class="navds-chips__chip-text"
                >
                  74 år
                </span>
              </button>
            </li>
            <li>
              <button
                aria-pressed="false"
                class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
              >
                <span
                  class="navds-chips__chip-text"
                >
                  75 år
                </span>
              </button>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="_container_21eff7 _container__hasPadding_21eff7"
    >
      <svg
        aria-labelledby="loader-r3i"
        aria-live="polite"
        class="navds-loader _loader__isCentered_1c4a2e navds-loader--3xlarge navds-loader--neutral"
        data-testid="alderspensjon-loader"
        focusable="false"
        preserveAspectRatio="xMidYMid"
        viewBox="0 0 50 50"
      >
        <title
          id="loader-r3i"
        >
          Et øyeblikk, vi beregner pensjonen din
        </title>
        <circle
          class="navds-loader__background"
          cx="25"
          cy="25"
          fill="none"
          r="20"
          xmlns="http://www.w3.org/2000/svg"
        />
        <circle
          class="navds-loader__foreground"
          cx="25"
          cy="25"
          fill="none"
          r="20"
          stroke-dasharray="50 155"
        />
      </svg>
      <h2
        class="navds-heading navds-heading--small"
      >
        Beregning
      </h2>
      <div
        class="_alert_9168cd"
        role="alert"
      >
        <div
          class="_alertWrapper_9168cd navds-body-long navds-body-long--medium"
        >
          Du har ikke høy nok opptjening til å kunne starte uttak ved 62 år. Prøv en høyere alder.
        </div>
      </div>
    </div>
  </div>
  <div
    class="_background_21eff7 _background__lightblue_21eff7"
  >
    <div
      class="_container_21eff7"
    >
      <section
        class="_section_095025"
      >
        <div
          class="_innerwrapper_095025"
        >
          <button
            class="_button_095025 navds-button navds-button--secondary navds-button--medium"
          >
            <span
              class="navds-label"
            >
              Tilbake til start
            </span>
          </button>
          <button
            class="_button_095025 _button__avbryt_095025 navds-button navds-button--tertiary navds-button--medium"
          >
            <span
              class="navds-label"
            >
              Avbryt
            </span>
          </button>
        </div>
      </section>
    </div>
  </div>
</DocumentFragment>
`;
